### PR TITLE
Add a Cassandra Queue V2 Implementation

### PR DESCRIPTION
**Pull request by MichaelSnowden**
**What changed?**
I defined the persistence.QueueV2 interface, which supersedes the persistence.Queue interface. I also added an implementation of this interface for Cassandra. 

The implementations for delete and merge are left out for now and will be implemented later. As a result, there is no logic to read/write from the queues table (to read/write ack levels). Later, the ReadMessages method will need to be updated to use this as the minimum message id.

**Why?**
We plan on using this for the upcoming history task DLQ project.

**How did you test it?**
The new code has 100% test coverage, even including a test for CAS application failures due to concurrent writers. I tested that particular one locally 100 times to verify that it wasn't flaky.

**Potential risks**
The biggest risk IMO is that because we're doing this piecemeal, we may realize there's an issue in the database schema and then have to change it later, which will be a pain if we need to ensure it's backwards compatible, but we can prevent this by ensuring that those changes are not split across multiple releases.

**Is hotfix candidate?**
No.

### CODE REVIEW

File: common/persistence/cassandra/queue_v2_store.go [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+	for iter.Scan(&messageID, &messagePayload, &messageEncoding) {
+		encoding, ok := enums.EncodingType_value[messageEncoding]
+		if !ok {
+			return nil, fmt.Errorf("%w: %v", ErrInvalidQueueMessageEncodingType, messageEncoding)
+		}
+
+		encodingType := enums.EncodingType(encoding)
+
+		// We need to clone the payload byte slice because its underlying array may be reused for another row.
+		payloadClone := slices.Clone(messagePayload)
```
Comment (yycptt) [inline common/persistence/cassandra/queue_v2_store.go:148]: create a new byte array for messagePayload on each iteration, then we can avoid the clone?
or check how existing queue impl does this, I don't think we are doing clone there.

File: common/persistence/cassandra/queue_v2_store.go [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+	ErrInvalidQueueMessageEncodingType = errors.New("invalid encoding type for queue message")
+)
+
+func NewQueueV2Store(session gocql.Session) persistence.QueueV2 {
+	return &queueV2Store{
+		session: session,
+	}
+}
+
+func (q *queueV2Store) EnqueueMessage(
```
Comment (yycptt) [inline common/persistence/cassandra/queue_v2_store.go:92]: we need to control the concurrency of this method. plz add a TODO comment.

File: common/persistence/cassandra/queue_v2_store.go [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+	//	schema/cassandra/temporal/versioned/v1.9/queues.cql
+	queueV2Store struct {
+		session gocql.Session
+	}
+)
+
+const (
+	TemplateEnqueueMessageQuery  = `INSERT INTO queue_messages (queue_type, queue_name, queue_partition, message_id, message_payload, message_encoding) VALUES (?, ?, ?, ?, ?, ?) IF NOT EXISTS`
+	TemplateGetMessagesQuery     = `SELECT message_id, message_payload, message_encoding FROM queue_messages WHERE queue_type = ? AND queue_name = ? AND queue_partition = ? AND message_id >= ? ORDER BY message_id ASC LIMIT ?`
+	TemplateGetMaxMessageIDQuery = `SELECT MAX(message_id) AS message_id FROM queue_messages WHERE queue_type = ? AND queue_name = ? AND queue_partition = ?`
```
[the remark this answers is not in the document]
Comment (MichaelSnowden) [reply common/persistence/cassandra/queue_v2_store.go]: Good catch--it looks like there actually is. I changed it to the latter.

File: common/persistence/cassandra/queue_v2_store.go [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+	// queue_messages tables that implement the QueueV2 interface. The schema is located at:
+	//	schema/cassandra/temporal/versioned/v1.9/queues.cql
+	queueV2Store struct {
+		session gocql.Session
+	}
+)
+
+const (
+	TemplateEnqueueMessageQuery  = `INSERT INTO queue_messages (queue_type, queue_name, queue_partition, message_id, message_payload, message_encoding) VALUES (?, ?, ?, ?, ?, ?) IF NOT EXISTS`
+	TemplateGetMessagesQuery     = `SELECT message_id, message_payload, message_encoding FROM queue_messages WHERE queue_type = ? AND queue_name = ? AND queue_partition = ? AND message_id >= ? ORDER BY message_id ASC LIMIT ?`
```
[the remark this answers is not in the document]
Comment (MichaelSnowden) [reply common/persistence/cassandra/queue_v2_store.go:50]: I want to keep just to be (make myself feel) safe

File: common/persistence/cassandra/queue_v2_store.go [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+		messageID,
+		blob.Data,
+		blob.EncodingType.String(),
+	).WithContext(ctx).MapScanCAS(make(map[string]interface{}))
+	if err != nil {
+		return gocql.ConvertError("QueueV2EnqueueMessage", err)
+	}
+
+	if !applied {
+		return fmt.Errorf("%w: insert with message ID %v was not applied", ErrQueueMessageConflict, messageID)
```
[the remark this answers is not in the document]
Comment (MichaelSnowden) [reply common/persistence/cassandra/queue_v2_store.go]: Fixed

File: common/persistence/cassandra/queue_v2_store.go [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+		Metadata: persistence.MessageMetadata{ID: messageID},
+	}, nil
+}
+
+func (q *queueV2Store) ReadMessages(
+	ctx context.Context,
+	request *persistence.InternalReadMessagesRequest,
+) (*persistence.InternalReadMessagesResponse, error) {
+	if request.PageSize <= 0 {
+		return nil, persistence.ErrNonPositiveReadQueueMessagesPageSize
```
[the remark this answers is not in the document]
Comment (MichaelSnowden) [reply common/persistence/cassandra/queue_v2_store.go:118]: Yes

File: common/persistence/cassandra/queue_v2_store.go [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+
+	return &persistence.InternalReadMessagesResponse{
+		Messages:      messages,
+		NextPageToken: nextPageToken,
+	}, nil
+}
+
+func (q *queueV2Store) getMinMessageID(request *persistence.InternalReadMessagesRequest) (int, error) {
+	if request.NextPageToken == nil {
+		return persistence.FirstQueueMessageID, nil
```
[the remark this answers is not in the document]
Comment (MichaelSnowden) [reply common/persistence/cassandra/queue_v2_store.go:180]: Added a TODO

File: common/persistence/cassandra/queue_v2_store.go (hunk 1 above)
Comment (MichaelSnowden) [reply common/persistence/cassandra/queue_v2_store.go]: Done

File: common/persistence/cassandra/queue_v2_store.go (hunk 2 above)
Comment (MichaelSnowden) [reply common/persistence/cassandra/queue_v2_store.go:93]: Added a TODO

**Review by yycptt — APPROVED** [review]

**Pull request merged**